Add basic parquet export capability to PudlSqliteIOManager

rousik opened the pull request:
This is an experimental support for emitting data in parquet format. It is currently controlled via env variables `PUDL_WRITE_TO_PARQUET` and `PUDL_READ_FROM_PARQUET` and both these modes are currently disabled.

rousik commented:
@zaneselvans once we actually support dual output formats (sqlite+parquet) it might be good time to do a little cleanup in the io manager codebase as this has been clearly designed with sqlite only in mind, and the class structure seems little clunky here, esp. when it comes to overlap/differences between pudl/ferc io managers (the latter being only partial and fixed to sqlite formats) and pudl/epacems (that might now share more due to both writing to parquet). I will think about how we could improve the situation, but this could get us going in the meantime.

zaneselvans commented on src/pudl/io_managers.py:
```diff
@@ -477,6 +509,21 @@ def _handle_pandas_output(self, context: OutputContext, df: pd.DataFrame):
+            logger.info(f"Writing {self.db_name}/{table_name} to parquet.")
+            parquet_path = PudlPaths().parquet_path(self.db_name, table_name)
+            parquet_path.parent.mkdir(parents=True, exist_ok=True)
+            schema = res.to_pyarrow()
```
Along with the schema, another resource-specific parameter we might want in here is information about how to organize the row groups for larger assets, which I guess could be another method we add to `Resource`

zaneselvans replied:
This doesn't currently exist. The only big table we've written out to Parquet historically is the EPA CEMS hourly emissions, and that's been done in a bespoke way.

rousik replied:
Yeah, I think we can encode this in the resource, or in the asset metadata once we're ready for it. The code should be flexible enough to accommodate that once we need it.

zaneselvans commented on src/pudl/io_managers.py:
```diff
@@ -510,6 +557,13 @@ def load_input(self, context: InputContext) -> pd.DataFrame:
+        if self.settings.read_from_parquet:
+            logger.info(f"Reading table {self.db_name}/{table_name} from parquet.")
+            schema = res.to_pyarrow()
+            return pq.read_table(
+                source=PudlPaths().parquet_path(self.db_name, table_name),
+                schema=schema,
+            ).to_pandas()
```
I don't think this pandas -> parquet -> pandas round-trip will necessarily preserve dtypes, since we're not using pyarrow dtypes in our dataframes, so I suspect we'll still need to`res.enforce_schema()` here as we do when reading out of SQLite below to avoid unexpected dtype problems downstream.

zaneselvans reviewed (changes requested):
I'm fuzzy on where the configuration **should** go, but I'm pretty sure that Dagster already has a place where it expects to receive this kind of resource configuration, including via environment variables. But maybe @bendnorman knows off the top of his head.

It feels like it would be cleaner to keep the SQLite IO Manager dedicated to SQLite, and have another IO Manager that is dedicated to the Parquet logic, and then compose them together into a hybrid that can do both / either depending on how it's configured. Does that seem reasonable?

rousik commented on src/pudl/io_managers.py:
```diff
@@ -382,12 +405,17 @@ class PudlSQLiteIOManager(SQLiteIOManager):
         db_name: str,
         package: Package | None = None,
         timeout: float = 1_000.0,
+        settings: ParquetSettings = ParquetSettings(),
```
I'm now attempting to switch to dagster configuration system, but it seems to me that `EnvVar` that dagster has is kind of crude compared to pydantic. They only support str/int types for env variables and AFAIK there's no easy way to say what the default should be when the env variable is not actually set. This seems really strange to me and I'm still investigating what the options are here.

rousik commented on src/pudl/io_managers.py:
```diff
@@ -382,12 +405,17 @@ class PudlSQLiteIOManager(SQLiteIOManager):
     using the :class:`pudl.metadata.classes.Package` class.
     """
 
+    # TODO(rousik): now that this experimentally supports also writing to parquet
+    # as an alternative storage format, we should probably rename this class
+    # to be less sqlite-centric.
+
```
I dug deeper into how we deal with io_managers and I'm a bit perplexed by the apparent complexity here. There are several partial implementations, some of which support reads, some of which only writes, and there's more duplicate code than ideally should be, but I think this is a problem for a little later and shouldn't be necessarily solved as part of this.

After some more tinkering, I think that the idea of composite IO manager that simply hands-off reads/writes to either sqlite or parquet should work well and should be relatively clean too.

The only remaining snag is that dagster support for loading configuration from env variables seems kind of... incomplete... in that there doesn't seem to be support for booleans or for supplying default value for when the env variable is not set. 

I was expecting something along the lines of 1. use built-in default, 2. replace this with env-variable if set, and 3. replace env default with custom value in case it is supplied via dagster ui/job configuration, but either I'm misreading or this is not actually supported?!

zaneselvans replied:
Among us I think @bendnorman knows the most about how Dagster resource configuration is supposed to work. I only have a vague understanding, and am just hoping that we can avoid a proliferation of configuration systems, and ideally use whatever the canonical system is for the framework we've adopted.

bendnorman replied:
Do we have to use an env var to configure the `PudlMixedFormatIOManager`? I think just using vanilla boolean attributes would be fine:

```python
write_to_parquet: bool = False
"""If true, data will be written to parquet files."""

read_from_parquet: bool = False
"""If true, data will be read from parquet files instead of sqlite."""
```

We could create a `etl_full_gcp_build` that configures the `PudlMixedFormatIOManager` to write and read from parquet.

You can also set a default value for a `EnvVar` using [`EnvVar.get_value()`](https://github.com/dagster-io/dagster/blob/9aa58e22e68e8455573cc04c7fb6a3448ce933c5/python_modules/dagster/dagster/_config/field_utils.py#L564C16-L568). We could do something like this:

```python
write_to_parquet: bool = bool(EnvVar("PUDL_WRITE_TO_PARQUET").get_value(default=True))
```

Not the most elegant thing ever but I tested it out and it allows you to have a default, read from an env var and overwrite the value in the UI. 

There is an [`IntEnvVar`](https://github.com/dagster-io/dagster/blob/9aa58e22e68e8455573cc04c7fb6a3448ce933c5/python_modules/dagster/dagster/_config/field_utils.py#L502) that casts env var values to ints. Maybe we could make one for booleans?

bendnorman replied:
I think my preference is for `PudlMixedFormatIOManager` to be a subclass of `IOManager` and to configure it using the `@io_manager(config_schema=...)` decorator. [These are the legacy concepts](https://docs.dagster.io/concepts/io-management/io-managers-legacy) but I don't think it's great to have our IO managers use a mix of the legacy and current resource concepts.

Eventually we should move to using the new pydantic configurable resources. We explored migrating our setting configurations in https://github.com/catalyst-cooperative/pudl/issues/2842 but decided not to because it was a bigger change than we expected. We could create separate issues for migrating the resources in `pudl.resources` and `pudl.io_managers`.

zaneselvans replied:
Ah okay, so there *is* a way to do all this configuration using the more ergonomic Pydantic settings system. But we would need to do some migration work to get the system there.

rousik replied:
@bendnorman `ConfigurableIOManager` is a subclass of `IOManager` and works just fine with the old-style `@io_manager` annotated method. The old-style approach just hands out instance of this io manager to whoever needs it, so I don't think there's a practical problem with how it's done now. Or maybe I'm missing something.

Using `ConfigurableIOManager` will AFAIK make the migration easier once we're ready for it because this is basically already prepared for the new world.

rousik replied:
This is fair point, the duplication of the config both in the `ConfigurableIOManager` and `config_schema` here would be tedious.

rousik replied:
@bendnorman, fair point re/ use of `config_schema` here. I will make the necessary changes.

rousik replied:
Though it's tempting to keep using `ConfigurableIOManager` as that takes care of building constructor for me that would be identical to what I'm about to write :-)

rousik commented:
Moving the conversation ahead a notch. The modified code that addresses some of the issues outlined here is still on my local machine and need to be cleaned up before publishing - this is pending open issue with dagster `EnvVar`.

rousik commented:
I've refactored IO managers so that there's sqlite, parquet and a combined one. Reading from env variables is kind of crummy, because dagster has much worse support for this than pydantic `BaseSettings`, but I don't think this is a major problem. parquet writes/reads will be enabled whenever the associated env variables are non-empty (i.e. truth value is based on strings, not interpreted as boolean, which could lead to some confusion).

I feel like there's still some work to be done to clean up things, and, ideally we would also use the same parquet io manager for epacems, but the presence of sharding will require some custom logic.

pyarrow parquet writer does support sharding, but the naming schema is usually `base_dir/key=value/{uuid}.parquet` i.e. in our case it would be `epacems/state=CA/...` which may not be what we want.

zaneselvans commented:
@rousik It looks like it's not actually able to write to the DB because it has no connection engine?

rousik commented:
Oh that would be tests reaching into the IO manager and assuming it's internal structure. I'll fix later today and when at the keyboard.

ETL itself should be working.

rousik commented on src/pudl/io_managers.py:
```diff
@@ -536,9 +638,19 @@ def load_input(self, context: InputContext) -> pd.DataFrame:
 
 @io_manager
-def pudl_sqlite_io_manager(init_context) -> PudlSQLiteIOManager:
+def pudl_sqlite_io_manager(init_context) -> IOManager:
```
@bendnorman and @zaneselvans - given that this now returns mixed format IO manager, could we rename this resource/io-manager from `pudl_sqlite_io_manager` to plain `pudl_io_manager`? I think that would make it more appropriate. There's a bit of a risk that very widely used word `pudl` refers both to *the primary output database* (`pudl.sqlite` up to this point) as well as the overall project/ETL/codebase. This ambiguity may be okay, because I'm not quite sure what would be the other word we could substitute here. Maybe `main_output` or something along those lines?

bendnorman replied:
I think `pudl_io_manager` is appropriate for this IO Manager. We're currently ok with the naming ambiguity of the code and data outputs. We explain the different parts of pudl in the [readme](https://github.com/catalyst-cooperative/pudl?tab=readme-ov-file#what-is-pudl).

zaneselvans replied:
I think `pudl_io_manager` makes sense.

This is probably beyond the scope of this PR and maybe more of a @bendnorman thing but I think we should also change the way the `default_resources` are defined in `pudl.etl.__init__.py`:

```py
default_resources = {
    "datastore": datastore,
    "pudl_sqlite_io_manager": pudl_sqlite_io_manager,
    "ferc1_dbf_sqlite_io_manager": ferc1_dbf_sqlite_io_manager,
    "ferc1_xbrl_sqlite_io_manager": ferc1_xbrl_sqlite_io_manager,
    "dataset_settings": dataset_settings,
    "ferc_to_sqlite_settings": ferc_to_sqlite_settings,
    "epacems_io_manager": epacems_io_manager,
}
```

It seems like the keys in this dictionary should not specify anything about the format / destination of the data, while the values (the actual IO managers) should have that specificity.  So `pudl_io_manager` could be any kind of IO manager... but regardless it's the one that's going to be used to output PUDL data, whether it's going to SQLite, BigQuery, both SQLite + Parquet, etc.  And the value associated with it determines the details.

I found this very confusing in thinking about how we were going to switch over to writing the data to not just SQLite, since we've hard-coded `pudl_sqlite_io_manager` all throughout the system.  But it's not really an SQLite IO Manager. It could be any kind of IO Manager.  But I didn't realize they were separable until reviewing this PR.

The FERC IO managers seem like they have a similar issue, and also, what IO Managers are used to output all of the FERC Forms that aren't Form 1?

bendnorman replied:
I agree Zane the keys shouldn't specify anything about format / destination. 

The FERC IO managers are just for reading data out of the dbf and xbrl sqlite databases. We don't process any of the other ferc forms in the main PUDL ETL so we don't have IO managers for them.

zaneselvans replied:
Also don't we need to tell it to use the new IO Manager here? Otherwise it won't actually generate any Parquet files will it?

rousik replied:
We are instantiating `PudlMixedFormatIOManager` here which uses either of the formats depending on the configuration.

zaneselvans commented on src/pudl/io_managers.py:
```diff
@@ -83,6 +87,72 @@ def __getitem__(self, idx):
+        )
+        self._parquet_io_manager = PudlParquetIOManager()
+
+        # TODO(rousik): Move the following diagnostics to debug level once
+        # we know what is happening.
+        write_fmt = "parquet" if self.write_to_parquet else "sqlite"
+        read_fmt = "parquet" if self.read_from_parquet else "sqlite"
+        logger.warning(
+            f"pudl_io_manager will be reading from {read_fmt}, writing to {write_fmt}."
+        )
```
There's an asymmetry here in that we can only read from one source, and it can be either SQLite or Parquet, but we're always writing to SQLite, and optionally also writing to Parquet. But the way it's phrased it seems like there's one read format, and one write format, which is misleading.

rousik replied:
Right, I have tweaked the log message. We probably want to drop the severity also. I have used `warning` for local testing to see that it correctly picks up env var overrides.

zaneselvans reviewed (changes requested):
Just one minor issue with what we're logging for the read/write formats.

zaneselvans commented:
Also, don't we need to update the resource configuration in `pudl/etl/__init__.py`? Otherwise it won't actually use the new IO manager will it?

rousik commented on src/pudl/io_managers.py:
```diff
@@ -83,6 +87,72 @@ def __getitem__(self, idx):
+    # and AFAIK there's no way to set default value if the env variable
+    # is not set.
+
+    write_to_parquet: bool = bool(EnvVar("PUDL_WRITE_TO_PARQUET").get_value(False))
+    """If true, data will be written to parquet files."""
+
+    read_from_parquet: bool = bool(EnvVar("PUDL_READ_FROM_PARQUET").get_value(False))
+    """If true, data will be read from parquet files instead of sqlite."""
```
In short, practical concerns.

I'm treating this as a "feature flag", where we start by having the new functionality (writing to parquet files) off by default, but toggle-able on demand. While toggling this in the dagster UI is useful for local development, in many cases where we might want to test this (e.g. in the CI, nightly builds and other dockerized/remote scenarios), it's very easy to set env variable and pass it to the ETL, and it's much more difficult to be tweaking dagster configuration from the outside IMO. 

This could be hooked into command-line flag, but that is also somewhat clunky and would require ad-hoc wiring to be piped through while env variables are easy and widely supported (both by github actions as well as when running docker image on vm/batch), so it should be fairly easy to toggle this new functionality just where we need it.

rousik replied:
If all goes well and this feature will end up being enabled everywhere and always, we can drop this toggle and hardcode the new behavior (always on). I'm unsure what would be the timeline for this, given the early stage this is in.

zaneselvans replied:
If reading from Parquet ends up significantly speeding up the ETL it's easy to imagine this getting turned on immediately.

Once it's there, I'd experiment with using it to speed up the integration tests and data validations (which would mean relying on SQLite primarily for bulk distribution, validation of referential integrity, and maybe other schema checks / constraints that are only implemented by the `to_sqlite()` schema). Since we want to remove `PudlTabl` (probably in the Spring) and that will mean reorganizing the tests, that would probably be a good time to play around with it. I don't know exactly what fraction of the time in the integration tests and data validations is spent reading data, but I suspect it's significant given the way we've wired up `PudlTabl` to pass the read requests through to the database and not cache anything in memory, and I think reading

I can't remember where we were discussing it but I think @bendnorman suggested that since the PUDL ETL script is now a relatively thin wrapper around some slightly nonstandard Dagster structures, we might want to switch to using the Dagster CLI directly, which I think would make these job configurations more straightforward with the Dagster configs.

rousik commented:
Okay. I have made some more changes, specifically renamed `pudl_sqlite_io_manager` to a generic `pudl_io_manager`. 

I have left `fake_pudl_sqlite_io_manager` in place, as that seems to be essentially unit tests for the `PudlSQLiteIOManager` itself. Personally I think that switching to `unittest.TestCase` here would make things a little cleaner instead of passing around these fixtures, e.g. (based on https://docs.python.org/3/library/unittest.html#basic-example)

```
class PudlSQLiteIOManagerTests(unittest.TestCase):

  def setUp(self):
    # spin up temp database (in TemporaryDirectory perhaps)
    self.io_manager = PudlSQLiteIOManager(...)

  def test_delete_statement(self):
    self.io_manager.handle_output(...)
    ...
```

We're already using `unittest` style tests for excel so that would not be first case :)

rousik commented on test/integration/etl_test.py:
```diff
@@ -33,7 +33,7 @@ def test_pudl_engine(
     if check_foreign_keys:
         # Raises ForeignKeyErrors if there are any
-        pudl_sql_io_manager.check_foreign_keys()
+        pudl_sql_io_manager._sqlite_io_manager.check_foreign_keys()
```
This is dirty, but I'm not sure what would be the best way around it. Maybe implement `MixedFormatIOManager.get_sqlite_io_manager() -> PudlSqliteIOManager` method that we could call when we really need to access sqlite only features?

bendnorman replied:
I think implementing a getter that throws a helpful error if the `MixedFormatIOManager` doesn't have a `sqlite_io_manager`.

It might not be worth the complexity but we could make the `MixedFormatIOManager` more generalizable and accept an arbitrary number of IO managers to use. Then we could have some methods for viewing and grabbing specific io managers:

```python
>>> MixedFormatIOManager.list_io_manager_names()
["sqlite_io_manager", "parquet_io_manager", "bq_io_manager", ...]
>>> MixedFormatIOManager.get_io_manager("sqlite_io_manager")
```

I think for now a simple `MixedFormatIOManager.get_sqlite_io_manager() -> PudlSqliteIOManager` method seems fine.

rousik replied:
I think the generalization here is probably not warranted. I would only add that if/when we know we will need it but otherwise it may be just future proofing for something that will never happen and I agree that the complexity is not worth it.

What this does show, however, is that sqlite and pudl-sqlite io managers are special in that their APIs and functionality is much broader than that of the typical/format-agnostic io manager and this API is a bit vague.

It might be valuable to separate these concerns somewhat, perhaps by defining the interface that such sqlite thingy should provide and have a method on the mixed io manager that can return implementations of this clearly-defined and separate interface. It could return the io_manager that implements this interface, but in most cases I see, we don't actually need the io_manager but the specific sqlite validation/schema-management functionality.

It might even make sense to separate this sqlite/schema functionality to a standalone class that is embedded in the IO manager but with which IO manager communicated through this well defined API.

Note that the above may very well go past the scope of this PR and should be likely reserved to subsequent cleanup/refactoring. Providing `get_sqlite_io_manager()` function is simple and deals with this dirtiness reasonably well.

rousik commented:
Looks like failing validation when there are unknown columns is not the right thing to do as demonstrated by the failing `ci-integration`. Perhaps the assumption in the test `test_extra_column_error` is itself wrong? Halp!

zaneselvans commented:
@rousik whenever you introduce additional frameworks or systems that we're not using (like `poetry` or `unittest`) it creates a barrier to our understanding, maintaining, and extending the work, or it means we need to dedicate additional internal resources to educating everyone on how these other systems work, or to replacing what you've done with the systems that we've already agreed to use internally.

bendnorman commented on src/pudl/io_managers.py:
```diff
@@ -570,6 +626,10 @@ def __init__(
                 connection opens a transaction to modify the database, it will be locked
                 until that transaction is committed.
         """
+        # TODO(rousik): Note that this is a bit of a partially implemented IO manager that
+        # is not actually used for writing anything. Given that this is derived from base
+        # SqliteIOManager, we do not support handling of parquet formats. This is probably
+        # okay for now.
```
Could you add these thoughts to the doc string? The doc string already states the class should be subclassed though it could be a little more clear.

bendnorman commented on test/unit/io_managers_test.py:
```diff
@@ -268,46 +99,174 @@ def test_migrations_match_metadata(tmp_path, monkeypatch):
-        fake_pudl_sqlite_io_manager_fixture.load_input(input_context)
+# TODO(rousik): is there difference between fake_sqlite_io_manager
+# and real sqlite_io_manager in terms of functionality?!!
+class PudlSQLiteIOManagerTest(unittest.TestCase):
```
Why are you using unittest instead of pytest here?

rousik replied:
See other threads on the PR. I thought that `unitest` offers generally cleaner tests with less verbosity. However, if this breaks the convention (most of the code does use `pytest` even though there are some `unitest` based tests too), I can revert these changes.

rousik replied:
I do like the ability to put all of the assumptions/expectations about single class-under-test into a parallel unittest test-class, as this provides nice encapsulation and visual separation compared to the flat undifferentiated pytest style approach, but if we think consistency is more valuable (pytest vs unittest) I can revert these last changes.

rousik commented:
> @rousik whenever you introduce additional frameworks or systems that we're not using (like `poetry` or `unittest`) it creates a barrier to our understanding, maintaining, and extending the work, or it means we need to dedicate additional internal resources to educating everyone on how these other systems work, or to replacing what you've done with the systems that we've already agreed to use internally.

This is a fair point. I agree that adding new technologies to the mix have non-trivial barriers and we shouldn't do this haphazardly, which is, perhaps, how it looks from the outside. Let me try to provide a bit more detailed reasoning for `unittest`:

1. `unittest` style is already used in excel, so we're not strictly introducing it from scratch. It was me who added it back then, so it's still on me though :)
2. `unittest` plays well in `pytest` in that pytest can run unittest based tests just fine, so it boils down to different "style" of writing tests, but the overall usage remains unchanged and unaffected.
3. It feels to me that unittest offers a little more encapsulated and nicer ways to build tests, which is particularly nice for situations where we are testing specific classes and their expectations (i.e. rather than having flat collection of tests, we can group them into test-classes that are geared towards specific functionality) and provide very explicit setup/teardown logic rather than relying on the nebulous cloud of fixtures.

It is, in particular, the point (3) where I think that departing from `pytest` might result in cleaner and leaner looking test and it might very well be worth it to use that. I do think that we're also using fairly limited and intuitive features from `unittest` so it shouldn't be too hard to maintain those tests, but perhaps I'm underestimating the burden, and overestimating the lower maintenance of these "nicer" tests.

I can roll these changes back if this is undesirable.

I think we can have separate conversation re/ `poetry` - there, the use is definitely more experimental, but also non-binding in that it should be possible to revert back/switch to other dep management systems.

rousik commented on src/pudl/etl/check_foreign_keys.py:
```diff
@@ -51,7 +51,12 @@ def pudl_check_fks(logfile: pathlib.Path, loglevel: str):
 
     context = build_init_resource_context()
-    io_manager = pudl_sqlite_io_manager(context)
+    io_manager: PudlSQLiteIOManager = pudl_io_manager(context)._sqlite_io_manager
+    # TODO(rousik): This reaches into the io_manager and assumes
+    # PudlSQLiteIOManager and the corresponding foreign key
+    # functionality. It's strange to be constructing dagster
+    # resources to achieve this, but alas, that's where the logic
+    # lies.
```
We have the foreign key check on the schemas/Packages (do the foreign keys make sense at all), and right now we have the validation on the actual data, which is very much format specific (relies on sqlite pragmas). Building something format agnostic may be well out of scope of this PR, but nevertheless, extracting the functionality out of io_manager does actually make sense and I will do so. That will simplify things.

zschira commented:
Hey @rousik, we decided we want to try to get the parquet support completed by the end of the week, so I'm going to branch off of here and see if I can get a parallel PR through over the finish line. Thanks for getting this most of the way there, though!

rousik commented:
> Hey @rousik, we decided we want to try to get the parquet support completed by the end of the week, so I'm going to branch off of here and see if I can get a parallel PR through over the finish line. Thanks for getting this most of the way there, though!

Thanks, that is awesome. I'm off the grid most of this week so it's a great decision.

zaneselvans commented:
Parquet outputs have been merrrrged in #3296
